### PR TITLE
drivers: serial: nrfx_uarte: Add support for read only TX buffers 

### LINKED ISSUE #40838: Nordic UART driver (UARTE) fail to transfer buffers from read only memory
**Describe the bug**
when `uart_tx` is called with buffer pointing outside of RAM it will return error. It is a hardware feature that EasyDMA can only transfer from RAM address but it should be handled by the driver with internal cache buffer.

**Impact**
When #40757 got merged asynchronous API can be used for logging. When logs are dropped, fixed message is sent and this message is not in RAM which results in error returned by `uart_tx` and eventually assert in UART logging backend.

### PR DESCRIPTION
**Pull request by nordic-krch**
UARTE does not support RO TX buffers. Added cache buffer to the driver which is used when provided buffer is not in RAM. In that case data is sent chunk by chunk. Size of the buffer is common for all instances and can be configured in Kconfig.

Extended tests to check RO buffers with `uart_tx`.

~~Due to error in k_timer #40391 needs to be cherry-picked to get tests to pass.~~

Fixes #40838.

### CODE REVIEW

**Comment by nordic-krch** [comment]
It would be great to get that one into 3.0.0 as it prevents us from using DMA transfers in UART logging backend on Nordic platforms (see #40757). @anangl any chance to look at it?

File: drivers/serial/uart_nrfx_uarte.c [hunk 1]
```diff
@@ -73,8 +73,12 @@ struct uarte_async_cb {
 	void *user_data;
 
 	const uint8_t *tx_buf;
+	const uint8_t *xfer_buf;
+	size_t xfer_len;
 	volatile size_t tx_size;
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte.c:78]: Nit: `tx_buf` and `tx_size` are coupled, so I think it's better to keep their definitions next to each other.

File: drivers/serial/uart_nrfx_uarte.c [hunk 2]
```diff
@@ -100,6 +104,7 @@ struct uarte_async_cb {
 	uint8_t rx_flush_cnt;
 	bool rx_enabled;
 	bool hw_rx_counting;
+	bool pend_tx;
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte.c:107]: Maybe `pending_tx` will be more clear?

File: drivers/serial/uart_nrfx_uarte.c [hunk 3]
```diff
@@ -735,17 +740,51 @@ static int uarte_nrfx_init(const struct device *dev)
 	return 0;
 }
 
+/* Attempt to start TX (asynchrnous transfer). If hardware is not ready then pending
+ * buffer is set. When current poll_out is completed pending transfer is started.
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte.c:744]:
```suggestion
/* Attempt to start TX (asynchronous transfer). If hardware is not ready, then pending
 * flag is set. When current poll_out is completed, pending transfer is started.
```

File: drivers/serial/Kconfig.nrfx [hunk 4]
```diff
@@ -17,6 +17,13 @@ menuconfig UART_NRFX
 
 if UART_NRFX
 
+config UART_ASYNC_TX_CACHE_SIZE
+	int "TX cache buffer size"
+	depends on UART_ASYNC_API
+	default 8
+	help
+	  TX cache buffer is used when provided buffer is read only.
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/serial/Kconfig.nrfx:25]: what about `TX buffer is not located in RAM`? Technically, we check if address is in RAM.

[the remark this answers is not in the document]
Comment (anangl) [reply drivers/serial/Kconfig.nrfx:25]: Works for me.

**Review by anangl — APPROVED** [review]

**Comment by nordic-krch** [comment]
@Mierunski @dcpleung Can anyone of you take a look? I would like to get it into v3.0.0

**Review by dcpleung — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**